cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp

```diff
@@ -0,0 +1,523 @@
+    read_state_ = std::make_unique<ParquetReadState>(read_sequence);
+}
+
+std::shared_ptr<arrow::ChunkedArray> VectorizedColumnReader::readBatch(int64_t batch_size)
+{
+    record_reader_->Reset();
+    record_reader_->Reserve(batch_size);
+
+    while (read_state_->hasMoreRead() && batch_size > 0)
+    {
+        const int64_t readNumber = read_state_->currentRead();
+        if (readNumber < 0)
+        {
+            const int64_t records_skipped = record_reader_->SkipRecords(-readNumber);
+            assert(records_skipped == -readNumber);
+            read_state_->skip(records_skipped);
+            assert(hasMoreRead());
+        }
+        else
+        {
+            const int64_t readBatch = std::min(batch_size, readNumber);
+            const int64_t records_read = record_reader_->ReadRecords(readBatch);
+            assert(records_read == readBatch);
+            batch_size -= records_read;
+            read_state_->read(records_read);
+            if (!read_state_->hasMoreRead())
+                nextRowGroup();
+        }
+    }
+
+    std::shared_ptr<arrow::ChunkedArray> result;
+    THROW_ARROW_NOT_OK(
+        parquet::arrow::TransferColumnData(record_reader_.get(), arrowField_, input_.descr(), local_engine::default_arrow_pool(), &result));
+    return result;
+}
+
+VectorizedParquetRecordReader::VectorizedParquetRecordReader(const DB::Block & header, const DB::FormatSettings & format_settings)
+    : format_settings_(format_settings)
+    , arrowColumnToCHColumn_(
+          header,
+          "Parquet",
+          format_settings.parquet.allow_missing_columns,
+          format_settings.null_as_default,
+          format_settings.date_time_overflow_behavior,
+          format_settings.parquet.case_insensitive_column_matching)
+{
+}
+
+bool VectorizedParquetRecordReader::initialize(
+    const DB::Block & header,
+    const std::shared_ptr<arrow::io::RandomAccessFile> & arrow_file,
+    const std::shared_ptr<ColumnIndexFilter> & column_index_filter,
+    const std::shared_ptr<parquet::FileMetaData> & metadata)
+{
+    auto file_reader = parquet::ParquetFileReader::Open(arrow_file, parquet::default_reader_properties(), metadata);
+    const parquet::ArrowReaderProperties properties;
+    const parquet::FileMetaData & file_metadata = *(file_reader->metadata());
+    const parquet::SchemaDescriptor * parquet_schema = file_metadata.schema();
+    const auto keyValueMetadata = file_metadata.key_value_metadata();
+    parquet::arrow::SchemaManifest manifest;
+    THROW_ARROW_NOT_OK(parquet::arrow::SchemaManifest::Make(parquet_schema, keyValueMetadata, properties, &manifest));
+    std::vector<std::shared_ptr<arrow::Field>> fields;
+    fields.reserve(manifest.schema_fields.size());
+    for (auto const & schema_field : manifest.schema_fields)
+        fields.emplace_back(schema_field.field);
+    const arrow::Schema schema(fields, keyValueMetadata);
+
+    /// column pruning
+    DB::ArrowFieldIndexUtil field_util(
```

Comment:
Make sure subcolumn pruning in tuple column works

Reply:
I don't support complex type in this PR